a3p-integration/proposals/p:upgrade-19/provisionPool.test.js

```diff
@@ -0,0 +1,230 @@
+ */
+const getProvisionedAddresses = async vstorageKit => {
+  const children = await vstorageKit.vstorage.keys('published.wallet');
+  return children;
+};
+
+const checkUserProvisioned = (addr, vstorageKit) =>
+  retryUntilCondition(
+    () => getProvisionedAddresses(vstorageKit),
+    children => children.includes(addr),
+    'Account not provisioned',
+    { maxRetries: 5, retryIntervalMs: 1000, log: console.log, setTimeout },
+  );
+
+test.before(async t => {
+  const vstorageKit = await makeVstorageKit(
+    { fetch },
+    { rpcAddrs: ['http://localhost:26657'], chainName: 'agoriclocal' },
+  );
+
+  t.context = {
+    vstorageKit,
+  };
+});
+
+test.serial('upgrade provisionPool', async t => {
+  await evalBundles(UPGRADE_PP_DIR);
+
+  const vatDetailsAfter = await getDetailsMatchingVats('provisionPool');
+  const { incarnation } = vatDetailsAfter.find(vat => vat.vatID === 'v28'); // provisionPool is v28
+
+  t.log(vatDetailsAfter);
+  t.is(incarnation, 1, 'incorrect incarnation');
+  t.pass();
+});
+
+test.serial(
+  `check provisionPool can recover purse and asset subscribers after upgrade`,
+  async t => {
+    // @ts-expect-error casting
+    const { vstorageKit } = t.context;
+
+    // Introduce new user then provision
+    const { address } = await introduceAndProvision('provisionTester');
+    await checkUserProvisioned(address, vstorageKit);
+
+    // Send USDC_axl to pp
+    const istBalanceBefore = await getISTBalance(PROVISIONING_POOL_ADDR);
+    await bankSend(PROVISIONING_POOL_ADDR, `500000${USDC_DENOM}`);
+
+    // Check IST balance
+    await waitUntilAccountFunded(
+      PROVISIONING_POOL_ADDR,
+      ambientAuthority,
+      { denom: 'uist', value: istBalanceBefore + 500000 },
+      { errorMessage: 'Provision pool not able to swap USDC_axl for IST.' },
+    );
+
+    // Introduce USD_LEMONS
+    await evalBundles(ADD_PSM_DIR);
+    await waitUntilContractDeployed('psm-IST-USD_LEMONS', ambientAuthority, {
+      errorMessage: 'psm-IST-USD_LEMONS instance not observed.',
+    });
+
+    // Provision the provisionPoolAddress. This is a workaround of provisionPoolAddress
+    // not having a depositFacet published to namesByAddress. Shouldn't be a problem since
+    // vat-bank keeps track of virtual purses per address basis. We need there to be
+    // depositFacet for provisionPoolAddress since we'll fund it with USD_LEMONS
+    await provision('provisionPoolAddress', PROVISIONING_POOL_ADDR);
+    await checkUserProvisioned(PROVISIONING_POOL_ADDR, vstorageKit);
+
+    // Send USD_LEMONS to provisionPoolAddress
+    const istBalanceBeforeLemonsSent = await getISTBalance(
+      PROVISIONING_POOL_ADDR,
+    );
+    await evalBundles(DEPOSIT_USD_LEMONS_DIR);
+
+    // Check balance again
+    await waitUntilAccountFunded(
+      PROVISIONING_POOL_ADDR,
+      ambientAuthority,
+      { denom: 'uist', value: istBalanceBeforeLemonsSent + 500000 },
+      { errorMessage: 'Provision pool not bale swap USDC_axl for IST.' },
```

Comment:
```suggestion
      { errorMessage: 'Provision pool not able to swap USDC_axl for IST.' },
```

Reply:
resolved in https://github.com/Agoric/agoric-sdk/commit/9c757093b0d5492e178362e5458e02e081cbd2b0